server/datastore/mysql/microsoft_mdm.go

```diff
@@ -90,3 +90,172 @@ func (ds *Datastore) MDMWindowsDeleteEnrolledDevice(ctx context.Context, mdmDevi
 
 	return ctxerr.Wrap(ctx, notFound("MDMWindowsEnrolledDevice"))
 }
+
+// MDMWindowsDeletePendingCommands deletes pending commands for a given device id.
+func (ds *Datastore) MDMWindowsDeletePendingCommands(ctx context.Context, deviceID string) error {
+	stmt := "DELETE FROM windows_mdm_pending_commands WHERE device_id = ?"
+
+	_, err := ds.writer(ctx).ExecContext(ctx, stmt, deviceID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "delete MDMWindowsDeletePendingCommands")
+	}
+
+	return nil
+}
+
+// MDMWindowsInsertPendingCommand inserts a new WindowMDMPendingCommand in the database
+func (ds *Datastore) MDMWindowsInsertPendingCommand(ctx context.Context, cmd *fleet.MDMWindowsPendingCommand) error {
+	stmt := `
+		INSERT INTO windows_mdm_pending_commands (
+		command_uuid,
+		device_id,
+		cmd_verb,
+		setting_uri,
+		setting_value,
+		data_type,
+		system_origin ) VALUES (?, ?, ?, ?, ?, ?, ?)
+	`
+	_, err := ds.writer(ctx).ExecContext(
+		ctx,
+		stmt,
+		cmd.CommandUUID,
+		cmd.DeviceID,
+		cmd.CmdVerb,
+		cmd.SettingURI,
+		cmd.SettingValue,
+		cmd.DataType,
+		cmd.SystemOrigin)
+	if err != nil {
+		if isDuplicate(err) {
+			return ctxerr.Wrap(ctx, alreadyExists("MDMWindowsPendingCommand", cmd.CommandUUID))
+		}
+		return ctxerr.Wrap(ctx, err, "inserting MDMWindowsPendingCommand")
+	}
+
+	return nil
+}
+
+// MDMWindowsGetAndRemovePendingCommands retrieves all commands for a given device ID from the windows_mdm_pending_commands table
+func (ds *Datastore) MDMWindowsGetAndRemovePendingCommands(ctx context.Context, deviceID string) ([]*fleet.MDMWindowsPendingCommand, error) {
```

Comment:
The concern I have with this is that if the caller fails to complete whatever it has to do with the pending commands, then it's impossible to retry because those commands have been deleted from the DB.

Reply:
That's a good point and I see what you mean. My initial idea was to implement the concept of a `command queue` here, where commands were removed as soon as they were processed. @mna do you think that having a flag that indicates if the pending command was processed should be the way to go?

Reply:
@marcosd4h Looking a bit more at the processing logic in the service layer, I think the pending commands are essentially those that exist in `windows_mdm_pending_commands` and that have no corresponding row yet in `windows_mdm_commands`, right? 

So if you don't delete them from the `pending` table (and I think it's fine to _not_ delete them for now, consistent with how the macos commands are stored AFAIK), the ones that are actually pending are those that are in `pending` without a matching row in `non-pending` (so to speak), and it would automatically properly return any commands that failed to be stored in `non-pending` without setting a flag.

My only remaining concern, however, would be that the place where you do insert in the `non-pending` table (`storeSyncMLCmd`) is done before returning those commands to the device (correct me if I'm wrong) and as such, we don't know for sure if the device did receive the commands. But that may be some edge case that we can handle differently later on (e.g. a cron that looks at non-pending commands that haven't received any response in a while, or the next time that device checks in, we look at any non-pending command that still haven't received a response and send it again, etc - thinking out loud).

Pinging @roperzh and @gillespi314 for validation/other ideas on this scenario.

Reply:
> Looking a bit more at the processing logic in the service layer, I think the pending commands are essentially those that exist in windows_mdm_pending_commands and that have no corresponding row yet in windows_mdm_commands, right?

That's correct. The `windows_mdm_pending_commands` is the command queue with pending device operations. These entries do not have "session" information as they have not been sent to the device yet.

When the device requests for server commands, the queued operations in `windows_mdm_pending_commands` get retrieved, and these operations are used to build the actual SyncML commands that will be sent to the device. These commands with MS-MDM sessionID, messageID, and cmdID information are stored in the `windows_mdm_commands` so their command response can be tracked.

> So if you don't delete them from the pending table (and I think it's fine to not delete them for now, consistent with how the macos commands are stored AFAIK), the ones that are actually pending are those that are in pending without a matching row in non-pending (so to speak), and it would automatically properly return any commands that failed to be stored in non-pending without setting a flag.

Got it. I think it would be good to follow the same approach. I'll see if it possible to put together logic to use the command GUID for this.

> My only remaining concern, however, would be that the place where you do insert in the non-pending table (storeSyncMLCmd) is done before returning those commands to the device (correct me if I'm wrong) and as such, we don't know for sure if the device did receive the commands. But that may be some edge case that we can handle differently later on (e.g. a cron that looks at non-pending commands that haven't received any response in a while, or the next time that device checks in, we look at any non-pending command that still haven't received a response and send it again, etc - thinking out loud).

We can determine if a sent command was not received if there was no response for it within the same MS-MDM session. Each SyncML protocol command in MS-MDM should receive a status response through the `Status` or `Results` verb within the same session. The backend logic can use this to determine when a given command needs to be retried/resent (The sessionID changed and there were sent commands for a previous sessionID that didn't not have  a response)

Reply:
Awesome, sounds good then. I think we can at least start with this and refine/iterate as needed if we find more edge cases we didn't think about.